# Informe EDA

In [234]:
%run lib.ipynb

# Lectura del dataset de películas
df_movies = pd.read_parquet("datasets/movies_dataset.parquet")
df_movies.info()
df_movies.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45376 entries, 0 to 45375
Data columns (total 20 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   belongs_to_collection  4488 non-null   object 
 1   budget                 45376 non-null  object 
 2   genres                 45376 non-null  object 
 3   id                     45376 non-null  object 
 4   original_language      45365 non-null  object 
 5   overview               44435 non-null  object 
 6   popularity             45376 non-null  float64
 7   production_companies   45376 non-null  object 
 8   production_countries   45376 non-null  object 
 9   release_date           45376 non-null  object 
 10  revenue                45376 non-null  float64
 11  runtime                45130 non-null  float64
 12  spoken_languages       45376 non-null  object 
 13  status                 45296 non-null  object 
 14  tagline                20398 non-null  object 
 15  ti

belongs_to_collection    budget  \
0  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1                                               None  65000000   
2  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3                                               None  16000000   
4  {'id': 96871, 'name': 'Father of the Bride Col...         0   

                                              genres     id original_language  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...    862                en   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   8844                en   
2  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...  15602                en   
3  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...  31357                en   
4                     [{'id': 35, 'name': 'Comedy'}]  11862                en   

                                            overview  popularity  \
0  Led by Woody, Andy's toys live happily in his ...   21.946943   
1  When siblings Judy and Peter discover an encha...   17.015539   
2  A family wedding reignites the ancient feud be...   11.712900   
3  Cheated on, mistreated and stepped on, the wom...    3.859495   
4  Just when George Banks has recovered from his ...    8.387519   

                                production_companies  \
0     [{'name': 'Pixar Animation Studios', 'id': 3}]   
1  [{'name': 'TriStar Pictures', 'id': 559}, {'na...   
2  [{'name': 'Warner Bros.', 'id': 6194}, {'name'...   
3  [{'name': 'Twentieth Century Fox Film Corporat...   
4  [{'name': 'Sandollar Productions', 'id': 5842}...   

                                production_countries release_date  \
0  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-10-30   
1  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-12-15   
2  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-12-22   
3  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-12-22   
4  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-02-10   

       revenue  runtime                                   spoken_languages  \
0  373554033.0     81.0           [{'iso_639_1': 'en', 'name': 'English'}]   
1  262797249.0    104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   
2          0.0    101.0           [{'iso_639_1': 'en', 'name': 'English'}]   
3   81452156.0    127.0           [{'iso_639_1': 'en', 'name': 'English'}]   
4   76578911.0    106.0           [{'iso_639_1': 'en', 'name': 'English'}]   

     status                                            tagline  \
0  Released                                               None   
1  Released          Roll the dice and unleash the excitement!   
2  Released  Still Yelling. Still Fighting. Still Ready for...   
3  Released  Friends are the people who let you be yourself...   
4  Released  Just When His World Is Back To Normal... He's ...   

                         title  vote_average  vote_count  release_year  \
0                    Toy Story           7.7      5415.0          1995   
1                      Jumanji           6.9      2413.0          1995   
2             Grumpier Old Men           6.5        92.0          1995   
3            Waiting to Exhale           6.1        34.0          1995   
4  Father of the Bride Part II           5.7       173.0          1995   

      return  
0  12.451801  
1   4.043035  
2   0.000000  
3   5.090760  
4   0.000000

## Chequeo de nulos

In [235]:
#suma de valores nulos por columna
print(columnas_de_nulos(df_movies))


belongs_to_collection    40888
original_language           11
overview                   941
runtime                    246
status                      80
tagline                  24978
dtype: int64


### De un total de 45376, la columna "belongs_to_collection" tiene un 90.1% de valores nulos, como además la columna no será utilizada en el API (ni en el recomendador), será eliminada

In [236]:
df_movies = df_movies.drop(['belongs_to_collection'], axis=1)
print(columnas_de_nulos(df_movies))

original_language       11
overview               941
runtime                246
status                  80
tagline              24978
dtype: int64


### Para la columna "original_language", tenemos 11 filas con valor nulo. 8 de ellos tienen sólo un valor en la columna "spoken_languages" así que lo imputaremos en "original_languaje", para las 4 restantes imputaremos el primer valor que se encuentre en #spoken_languajes". Esta columna es de interes ya que posteriormente la utilizaré para reducir el dataset y poder factibilizar la ejecución del algoritmo de recomendación, ya que con todo el dataset genera problemas de memoria.

In [237]:
df_movies.loc[[21579, 22805, 32103], ['original_language']] = 'en'
df_movies.loc[[19559, 44494], ['original_language']] = ''
df_movies.loc[[37354], ['original_language']] = 'cs'
df_movies.loc[[40982], ['original_language']] = 'ur'
df_movies.loc[[41802, 44573], ['original_language']] = 'xx'
df_movies.loc[[43978], ['original_language']] = 'fr'
df_movies.loc[[44328], ['original_language']] = 'sv'

print(columnas_de_nulos(df_movies))


overview      941
runtime       246
status         80
tagline     24978
dtype: int64


### Utilizaré la columna "overview" como la base del algoritmo de recomendación, por lo que necesito que este poblada para todas las películas. Utilizaré en primera instancia la columna "tagline" y posteriormente la columna "title".

In [238]:

#Reemplazo los valores nulos de la columna overview por su correspondiente en tagline
df_movies["overview"] = df_movies["overview"].fillna(df_movies["tagline"])

#Reemplazo los valores nulos de la columna overview por su correspondiente en title
df_movies["overview"] = df_movies["overview"].fillna(df_movies["title"])

print(columnas_de_nulos(df_movies))


runtime      246
status        80
tagline    24978
dtype: int64


### La columna "runtime" es de interes para generar correlaciones posteriormente, voy a utilizar el valor promedio para rellenar los valores nulos ya que representan sólo el 0.51% de los datos en esa columna. 

In [239]:
df_movies["runtime"] = df_movies["runtime"].fillna(df_movies["runtime"].mean())
print(columnas_de_nulos(df_movies))


status        80
tagline    24978
dtype: int64


### Todos los registros con valor nulo en la columna "status", presentan "release_year" en fechas pasadas y además indicadores en la columna "popularity", por loq ue asignaremos el valor "Released" para todos los registros con valor nulo.

In [240]:
df_movies[df_movies['status'].isnull()]

budget                                             genres      id  \
189        0                [{'id': 99, 'name': 'Documentary'}]   56088   
765        0                [{'id': 99, 'name': 'Documentary'}]   52059   
1382       0                     [{'id': 35, 'name': 'Comedy'}]   40506   
2138       0                                                 []   77314   
4126       0  [{'id': 18, 'name': 'Drama'}, {'id': 53, 'name...  109472   
...      ...                                                ...     ...   
44295      0                                                 []   39827   
44983      0  [{'id': 18, 'name': 'Drama'}, {'id': 10749, 'n...   49492   
45005      0                                                 []  128268   
45073      0                                                 []   41811   
45353      0                                                 []   44330   

      original_language                                           overview  \
189                  en  A documentary about the culture of hip-hop. Th...   
765                  en  The Gate of Heavenly Peace is a feature-length...   
1382                 en  Wally Sparks is a tabloid TV show reporter who...   
2138                 fr                         The Cabinet of Dr. Ramirez   
4126                 en                                           The Girl   
...                 ...                                                ...   
44295                en              A celebration of 30 years of The Who.   
44983                en  Jogwa actually means alms given to a person, u...   
45005                en         Documentary about Veruschka von Lehndorff.   
45073                en  At a spacious house in Los Angeles, Helene is ...   
45353                en  A combination gambling den and bawdy house is ...   

       popularity production_companies  \
189      0.159596                   []   
765      0.107050                   []   
1382     0.576741                   []   
2138     0.037500                   []   
4126     0.001653                   []   
...           ...                  ...   
44295    0.014260                   []   
44983    0.004990                   []   
45005    0.056298                   []   
45073    0.000002                   []   
45353    0.219260                   []   

                                    production_countries release_date  \
189                                                   []   1995-08-25   
765                                                   []   1995-01-01   
1382   [{'iso_3166_1': 'US', 'name': 'United States o...   1997-01-01   
2138                                                  []   1991-12-04   
4126                                                  []   2001-06-06   
...                                                  ...          ...   
44295                                                 []   1994-08-15   
44983                                                 []   2009-09-25   
45005          [{'iso_3166_1': 'DE', 'name': 'Germany'}]   2005-09-30   
45073  [{'iso_3166_1': 'US', 'name': 'United States o...   1990-11-30   
45353           [{'iso_3166_1': 'FR', 'name': 'France'}]   1905-01-01   

       revenue     runtime                                   spoken_languages  \
189        0.0   93.000000                                                 []   
765        0.0  180.000000                                                 []   
1382       0.0  105.000000           [{'iso_639_1': 'en', 'name': 'English'}]   
2138       0.0   94.181675                                                 []   
4126       0.0    0.000000                                                 []   
...        ...         ...                                                ...   
44295      0.0  155.000000                                                 []   
44983      0.0   92.000000                                                 []   
45005      0.0   52.000000  [{'iso_639_1': 'de', 'name': 

In [241]:
df_movies['status'] = df_movies['status'].fillna('Released')
print(columnas_de_nulos(df_movies))

tagline    24978
dtype: int64


### La columna "tagline" presenta un 50% de valores nulos, así que sólo le asignaremos ''.

In [242]:
df_movies["tagline"] = df_movies["tagline"].fillna('')
print(columnas_de_nulos(df_movies))
print('\n')
df_movies.info()

Series([], dtype: float64)


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45376 entries, 0 to 45375
Data columns (total 19 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   budget                45376 non-null  object 
 1   genres                45376 non-null  object 
 2   id                    45376 non-null  object 
 3   original_language     45376 non-null  object 
 4   overview              45376 non-null  object 
 5   popularity            45376 non-null  float64
 6   production_companies  45376 non-null  object 
 7   production_countries  45376 non-null  object 
 8   release_date          45376 non-null  object 
 9   revenue               45376 non-null  float64
 10  runtime               45376 non-null  float64
 11  spoken_languages      45376 non-null  object 
 12  status                45376 non-null  object 
 13  tagline               45376 non-null  object 
 14  title                 45376 non-null  obj

## Valores duplicados

### Buscaremos valores duplicados para filas completas

In [243]:
df_movies[df_movies.duplicated(keep=False)]

budget                                             genres      id  \
676         0  [{'id': 18, 'name': 'Drama'}, {'id': 10749, 'n...  105045   
1463        0  [{'id': 18, 'name': 'Drama'}, {'id': 10749, 'n...  105045   
7340        0  [{'id': 80, 'name': 'Crime'}, {'id': 18, 'name...    5511   
8063        0  [{'id': 12, 'name': 'Adventure'}, {'id': 16, '...   23305   
9160        0  [{'id': 80, 'name': 'Crime'}, {'id': 18, 'name...    5511   
9322        0  [{'id': 12, 'name': 'Adventure'}, {'id': 16, '...   23305   
13215       0                [{'id': 99, 'name': 'Documentary'}]  159849   
13256       0  [{'id': 53, 'name': 'Thriller'}, {'id': 9648, ...  141971   
13370       0  [{'id': 53, 'name': 'Thriller'}, {'id': 9648, ...  141971   
13598       0  [{'id': 18, 'name': 'Drama'}, {'id': 35, 'name...  168538   
13940       0                     [{'id': 35, 'name': 'Comedy'}]   97995   
13994       0  [{'id': 28, 'name': 'Action'}, {'id': 27, 'nam...   18440   
14006       0  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   11115   
15695       0  [{'id': 18, 'name': 'Drama'}, {'id': 10769, 'n...   42495   
16160       0  [{'id': 28, 'name': 'Action'}, {'id': 18, 'nam...   99080   
16757       0  [{'id': 53, 'name': 'Thriller'}, {'id': 9648, ...  141971   
17220       0                      [{'id': 18, 'name': 'Drama'}]   25541   
19873       0  [{'id': 14, 'name': 'Fantasy'}, {'id': 18, 'na...  119916   
19908       0  [{'id': 18, 'name': 'Drama'}, {'id': 878, 'nam...  152795   
21143       0  [{'id': 14, 'name': 'Fantasy'}, {'id': 18, 'na...  119916   
21830       0  [{'id': 18, 'name': 'Drama'}, {'id': 878, 'nam...  152795   
22125       0  [{'id': 28, 'name': 'Action'}, {'id': 27, 'nam...   18440   
23017       0                      [{'id': 18, 'name': 'Drama'}]   25541   
24135       0  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...  265189   
24814       0  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   11115   
28829       0  [{'id': 18, 'name': 'Drama'}, {'id': 35, 'name...  168538   
29343       0  [{'id': 18, 'name': 'Drama'}, {'id': 10769, 'n...   42495   
33144  980000  [{'id': 18, 'name': 'Drama'}, {'id': 14, 'name...  298721   
35751       0                [{'id': 99, 'name': 'Documentary'}]  159849   
38813       0  [{'id': 28, 'name': 'Action'}, {'id': 18, 'nam...   99080   
39980  980000  [{'id': 18, 'name': 'Drama'}, {'id': 14, 'name...  298721   
40213       0                     [{'id': 35, 'name': 'Comedy'}]   97995   
45178       0  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...  265189   

      original_language                                           overview  \
676                  de  East-Berlin, 1961, shortly after the erection ...   
1463                 de  East-Berlin, 1961, shortly after the erection ...   
7340                 fr  Hitman Jef Costello is a perfectionist who alw...   
8063                 en  In feudal India, a warrior (Khan) who renounce...   
9160                 fr  Hitman Jef Costello is a perfectionist who alw...   
9322                 en  In feudal India, a warrior (Khan) who renounce...   
13215                en  The third film of Frank Capra's 'Why We Fight"...   
13256                fi  Recovering from a nail gun shot to the head an...   
13370                fi  Recovering from a nail gun shot to the head an...   
13598                en  In Zola's Paris, an ingenue arrives at a tony ...   
13940                en  After breaking a mirror in his home, superstit...   
13994                en  When a comet strikes Earth and kicks up a clou...   
14006                en  As an ex-gambler teaches a hot-shot college ki...   
15695                en  King Lear, old and tired, divides his kingdom ...   
16160                en  Originally called White Thunder, American prod...   
16757                fi  Recovering from a nail gun shot to the head an...   
17220                da  Former Danish servicemen Lars and Jimmy are th...   
19873             

### Eliminaremos las duplicadas quedándonos con la primera ocurrencia

In [244]:
df_movies = df_movies.drop_duplicates(keep='first')
df_movies[df_movies.duplicated(keep=False)]


Empty DataFrame
Columns: [budget, genres, id, original_language, overview, popularity, production_companies, production_countries, release_date, revenue, runtime, spoken_languages, status, tagline, title, vote_average, vote_count, release_year, return]
Index: []

### Buscaremos valores duplicados sólo para la columna id

In [245]:
df_movies[df_movies.duplicated(keep=False, subset=['id'])].sort_values(by=['id','popularity'])

budget                                             genres      id  \
44738  16000000  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   10991   
4110   16000000  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   10991   
20877         0                      [{'id': 18, 'name': 'Drama'}]  109962   
5706          0                      [{'id': 18, 'name': 'Drama'}]  109962   
23507   3512454                      [{'id': 18, 'name': 'Drama'}]  110428   
4352    3512454                      [{'id': 18, 'name': 'Drama'}]  110428   
44743         0  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   12600   
5531          0  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   12600   
11337      2500  [{'id': 18, 'name': 'Drama'}, {'id': 35, 'name...   13209   
15758      2500  [{'id': 18, 'name': 'Drama'}, {'id': 35, 'name...   13209   
836           0                      [{'id': 18, 'name': 'Drama'}]  132641   
29968         0                      [{'id': 18, 'name': 'Drama'}]  132641   
12061   1600000  [{'id': 18, 'name': 'Drama'}, {'id': 80, 'name...   14788   
10414   1600000  [{'id': 18, 'name': 'Drama'}, {'id': 80, 'name...   14788   
33702  26000000  [{'id': 12, 'name': 'Adventure'}, {'id': 10751...   15028   
5126   26000000  [{'id': 12, 'name': 'Adventure'}, {'id': 10751...   15028   
947           4  [{'id': 18, 'name': 'Drama'}, {'id': 10749, 'n...   22649   
15067         4  [{'id': 18, 'name': 'Drama'}, {'id': 10749, 'n...   22649   
33785  30000000  [{'id': 35, 'name': 'Comedy'}, {'id': 80, 'nam...    4912   
5861   30000000  [{'id': 35, 'name': 'Comedy'}, {'id': 80, 'nam...    4912   
26595  10000000  [{'id': 18, 'name': 'Drama'}, {'id': 27, 'name...   69234   
9571   10000000  [{'id': 18, 'name': 'Drama'}, {'id': 27, 'name...   69234   
20821  40000000  [{'id': 12, 'name': 'Adventure'}, {'id': 18, '...   77221   
11150  40000000  [{'id': 12, 'name': 'Adventure'}, {'id': 18, '...   77221   
2562          0                [{'id': 99, 'name': 'Documentary'}]   84198   
21094         0                [{'id': 99, 'name': 'Documentary'}]   84198   

      original_language                                           overview  \
44738                ja  When Molly Hale's sadness of her father's disa...   
4110                 ja  When Molly Hale's sadness of her father's disa...   
20877                en  Two literary women compete for 20 years: one w...   
5706                 en  Two literary women compete for 20 years: one w...   
23507                fr  Winter, 1915. Confined by her family to an asy...   
4352                 fr  Winter, 1915. Confined by her family to an asy...   
44743                ja  All your favorite Pokémon characters are back,...   
5531                 ja  All your favorite Pokémon characters are back,...   
11337                fa  Since women are banned from soccer matches, Ir...   
15758                fa  Since women are banned from soccer matches, Ir...   
836                  ja  Ten years into a marriage, the wife is disappo...   
29968                ja  Ten years into a marriage, the wife is disappo...   
12061                en  Set against the backdrop of a decaying Midwest...   
10414                en  Set against the backdrop of a decaying Midwest...   
33702                en  Until now, Zak Gibbs' greatest challenge has b...   
5126                 en  Until now, Zak Gibbs' greatest challenge has b...   
947                  en  British nurse Catherine Barkley (Helen Hayes) ...   
15067                en  British nurse Catherine Barkley (Helen Hayes) ...   
33785                en  Television made him famous, but his biggest hi...   
5861                 en  Television made him famous, but his biggest hi...   
26595                en  Count de Chagnie has discovered Christine's si...   
9571                 en  Count de Chagnie has discovered Christine's si...   
20821                en  On the Arabian Peninsula in the 1930s, two war...   
11150                en  On th

### Existen valores duplicados en la columna "id", la diferencia al analizar los registros se concentra en la columna "popularity", eliminaremos los registros duplicados conservando el que tenga la mayor popularidad.

In [ ]:
df_movies = df_movies.sort_values(by=['id','popularity'], kind='mergesort').drop_duplicated('id', keep='last')
#df_movies[df_movies.duplicated(keep=False, subset=['id'])].sort_values(by=['id','popularity'])
df_movies.info()

/tmp/ipykernel_10204/1345200008.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  df_movies.sort_values(by=['id','popularity'])[df_movies.duplicated(keep="first", subset=['id'])].drop()


ValueError: Need to specify at least one of 'labels', 'index' or 'columns'

## Valores faltantes

### Realizaremos el análisis para las columnas relevantes para el funcionamiento de los endpoints de la API. Estos son las columnas "id", "original_language", "popularity", "title", "vote_average", "vote_count", "return", "overview" y "release_date".

In [ ]:
selected_columns = ['id','original_language','title','release_date','return','vote_count','vote_average','popularity','overview']
print(df_movies[selected_columns].isna().sum())


id                   0
original_language    0
title                0
release_date         0
return               0
vote_count           0
vote_average         0
popularity           0
overview             0
dtype: int64


### No se detectan valores faltantes.

In [ ]:
selected_columns = ['id','original_language','title','release_date','return','vote_count','vote_average','popularity','overview']
df_movies.info()

<class 'pandas.core.frame.DataFrame'>
Index: 26 entries, 44738 to 21094
Data columns (total 19 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   budget                26 non-null     object 
 1   genres                26 non-null     object 
 2   id                    26 non-null     object 
 3   original_language     26 non-null     object 
 4   overview              26 non-null     object 
 5   popularity            26 non-null     float64
 6   production_companies  26 non-null     object 
 7   production_countries  26 non-null     object 
 8   release_date          26 non-null     object 
 9   revenue               26 non-null     float64
 10  runtime               26 non-null     float64
 11  spoken_languages      26 non-null     object 
 12  status                26 non-null     object 
 13  tagline               26 non-null     object 
 14  title                 26 non-null     object 
 15  vote_average          2

## Tipos de datos

### se analizarán los tipos de datos para las columnas relevantes.

## Análisis estadístico

## Análisis de correlaciones